File: client/src/utils/geoUtils.js
// src/utils/geoUtils.js

/**
 * Calculates the distance between two GPS coordinates in miles using the Haversine formula.
 * @param {{lat: number, lng: number}} p1 - The first point.
 * @param {{lat: number, lng: number}} p2 - The second point.
 * @returns {number} The distance in miles.
 */
export const haversineDistance = (p1, p2) => {
  if (!p1 || !p2) return 0;
  const R = 3958.8; // Earth's radius in miles
  const rad = (x) => x * Math.PI / 180;
  const dLat = rad(p2.lat - p1.lat);
  const dLon = rad(p2.lng - p1.lng);
  const lat1 = rad(p1.lat);
  const lat2 = rad(p2.lat);
  const a = Math.sin(dLat / 2) * Math.sin(dLat / 2) + Math.sin(dLon / 2) * Math.sin(dLon / 2) * Math.cos(lat1) * Math.cos(lat2);
  const c = 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));
  return R * c;
};

/**
 * Processes a raw route path to calculate cumulative distance and simulate elevation.
 * This function creates the data structure needed for the elevation profile chart.
 * Used by the mock API to generate realistic test data.
 * @param {Array<{lat: number, lng: number}>} path - An array of coordinate points.
 * @returns {Array<{distance: number, elevation: number}>} An array of data points for charting.
 */
export const processRouteForElevation = (path) => {
  if (!path || path.length < 2) return [];

  let cumulativeDistance = 0;
  let currentElevation = 500; // Starting elevation in feet
  
  const profileData = [{ distance: 0, elevation: currentElevation }];

  for (let i = 1; i < path.length; i++) {
    const segmentDistance = haversineDistance(path[i - 1], path[i]);
    cumulativeDistance += segmentDistance;

    // Simulate elevation change: add a random value between -20 and +20 feet
    currentElevation += (Math.random() - 0.5) * 40;
    // Ensure elevation doesn't go below a certain floor
    if (currentElevation < 300) currentElevation = 300;

    profileData.push({
      distance: cumulativeDistance,
      elevation: currentElevation,
    });
  }

  return profileData;
};

/**
 * Parses a KML text string into an XML Document.
 * @param {string} kmlText The raw string content of a KML file.
 * @returns {XMLDocument} A parsed XML document.
 */
export const parseKML = (kmlText) => {
  if (!kmlText || typeof kmlText !== 'string') {
    kmlText = '<?xml version="1.0" encoding="UTF-8"?>';
  }
  const parser = new DOMParser();
  return parser.parseFromString(kmlText, 'text/xml');
};

/**
 * Extracts an array of {lat, lng} coordinate objects from KML text.
 * @param {string} kmlText The raw string content of a KML file.
 * @returns {Array<{lat: number, lng: number}>|null} An array of coordinate objects.
 */
export const getCoordsFromKML = (kmlText) => {
  if (!kmlText) return null;
  const kmlDoc = parseKML(kmlText);
  const coordinatesNode = kmlDoc.querySelector('coordinates');
  if (!coordinatesNode) {
    console.error("No <coordinates> tag found in KML file.");
    return null;
  }
  const coordinatesString = coordinatesNode.textContent;
  return coordinatesString.trim().split(/\s+/).map(coordString => {
    const [longitude, latitude] = coordString.split(',');
    return { lat: parseFloat(latitude), lng: parseFloat(longitude) };
  }).filter(loc => !isNaN(loc.lat) && !isNaN(loc.lng));
};

File: server/utils/geo_utils.py
# utils/geo_utils.py
from math import radians, sin, cos, sqrt, atan2, asin, degrees

# Earth's radius in miles
EARTH_RADIUS_MILES = 3958.8

def get_destination_point(lat, lon, bearing, distance_miles):
    """
    Calculates a new GPS coordinate given a starting point, bearing, and distance.
    """
    lat_rad = radians(lat)
    lon_rad = radians(lon)
    bearing_rad = radians(bearing)

    lat2_rad = asin(sin(lat_rad) * cos(distance_miles / EARTH_RADIUS_MILES) +
                    cos(lat_rad) * sin(distance_miles / EARTH_RADIUS_MILES) * cos(bearing_rad))

    lon2_rad = lon_rad + atan2(sin(bearing_rad) * sin(distance_miles / EARTH_RADIUS_MILES) * cos(lat_rad),
                               cos(distance_miles / EARTH_RADIUS_MILES) - sin(lat_rad) * sin(lat2_rad))

    return degrees(lat2_rad), degrees(lon2_rad)

def get_bounding_box(origin_lat, origin_lon, total_radius_miles):
    """
    Calculates a square bounding box around a central point.

    Args:
        origin_lat (float): Latitude of the center point.
        origin_lon (float): Longitude of the center point.
        total_radius_miles (float): The total radius (e.g., search radius + path distance) in miles.

    Returns:
        tuple: A tuple of (min_lat, min_lon, max_lat, max_lon).
    """
    # Calculate the four corner points of the box using the total radius
    north_lat, _ = get_destination_point(origin_lat, origin_lon, 0, total_radius_miles)
    _, east_lon = get_destination_point(origin_lat, origin_lon, 90, total_radius_miles)
    south_lat, _ = get_destination_point(origin_lat, origin_lon, 180, total_radius_miles)
    _, west_lon = get_destination_point(origin_lat, origin_lon, 270, total_radius_miles)

    return south_lat, west_lon, north_lat, east_lon

def haversine_distance(p1, p2):
    """Calculates the distance between two GPS coordinates in miles."""
    lat1, lon1 = radians(p1['lat']), radians(p1['lon'])
    lat2, lon2 = radians(p2['lat']), radians(p2['lon'])
    dlon = lon2 - lon1
    dlat = lat2 - lat1
    a = sin(dlat / 2)**2 + cos(lat1) * cos(lat2) * sin(dlon / 2)**2
    c = 2 * atan2(sqrt(a), sqrt(1 - a))
    return EARTH_RADIUS_MILES * c

def interpolate_point(p1, p2, fraction):
    """Calculates an intermediate GPS point between two points."""
    lat1, lon1 = radians(p1['lat']), radians(p1['lon'])
    lat2, lon2 = radians(p2['lat']), radians(p2['lon'])
    d = 2 * asin(sqrt(sin((lat1 - lat2) / 2)**2 + cos(lat1) * cos(lat2) * sin((lon1 - lon2) / 2)**2))
    A = sin((1 - fraction) * d) / sin(d)
    B = sin(fraction * d) / sin(d)
    x = A * cos(lat1) * cos(lon1) + B * cos(lat2) * cos(lon2)
    y = A * cos(lat1) * sin(lon1) + B * cos(lat2) * sin(lon2)
    z = A * sin(lat1) + B * sin(lat2)
    lat = atan2(z, sqrt(x**2 + y**2))
    lon = atan2(y, x)
    return {'lat': degrees(lat), 'lon': degrees(lon)}